Actuator: Change return type to be a function of the UAVO

kubark42 opened the pull request:
The MixerSettings object variable type can change, so this lets the actuator code reflect that.

mlyle commented on flight/Modules/Actuator/actuator.c:
```diff
@@ -338,7 +338,7 @@ static void actuator_task(void* parameters)
 		ActuatorDesiredData *desired)
 {
-	int8_t *vector = get_mixer_vec(index);
+	typeof(&(*mixerSettings.Mixer1Vector)) vector = get_mixer_vec(index);
```
consider for these `typeof (mixerSettings.Mixer1Vector) (*vector)` .  The dereferences below would need `(*vector)[MIXERSETTINGS ...`.  The upside is that this form (pointer to array) preserves type information so smart compilers (clang, some versions of gcc) can detect accesses past the end.

note that pointer to array and pointer to first element of an array have the same value, so the `(*vector)` in access is kinda.. vestigial.

kubark42 replied:
Ugh, torn between having the compiler do smart things for us and having the code be readable. 

`(*vector)[0]` isn't a pattern we see too often. I suppose I could just put a comment in the code to make it clear what's going on. Maybe that'll be good enough so underexperienced devs don't get hung up on this.

kubark42 replied:
Actually, @mlyle, I'm not quite sure what you're suggesting. The functions can't return `typeof(mixerSettings.Mixer1Vector)` because it's an array type. So are you suggesting that only at the point of L341 that be switched to `typeof (mixerSettings.Mixer1Vector) (*vector)  = get_mixer_vec(index);`? I don't think this works, and the compiler seems to agree.

mlyle replied:
Here's an example:

https://gist.github.com/mlyle/2fc29501dd7ad15ba2a6

During compilation it prints out the warnings you'd want:

```
jar% gcc -o example example.c -O2 -Wall
example.c: In function ‘myFunc2’:
example.c:13:2: warning: return from incompatible pointer type [enabled by default]
  return &baaz;
  ^
example.c: In function ‘main’:
example.c:27:26: warning: array subscript is above array bounds [-Warray-bounds]
  printf("%d\n", (*vector)[16]);
```

kubark42 replied:
Gotcha. Fixed as suggested.

kubark42 committed:
Actuator: Change return type to be a function of the UAVO

The MixerSettings object variable size can change, so this lets the code
reflect that. It accomplishes this by taking the pointer to the array and
then deferencing it. While this might look weird, taking the pointer to the
array preserves type information so smart compilers can detect accesses
past the end.

mlyle commented:
LGTM.  Has it been tested?

kubark42 commented:
Yes, flight tested and works as expected.

Works well at d-ronin, so merging here as well.
